src/ds3.c

```diff
@@ -960,44 +964,49 @@ static ds3_owner* _parse_owner(const ds3_log* log, const xmlDocPtr doc, const xm
         }
     }
 
-    return owner;
+    *_response = owner;
+
+    return NULL;
 }
 
-static ds3_get_service_response* _parse_get_service_response(const ds3_log* log, const xmlDocPtr doc, const xmlNodePtr root) {
-    ds3_get_service_response* response = g_new0(ds3_get_service_response, 1);
+static ds3_error* _parse_get_service_response(const ds3_client* client, const ds3_request* request, ds3_get_service_response** _response) {
+    xmlDocPtr doc;
+    xmlNodePtr root;
+    ds3_error* error;
+    ds3_get_service_response* response;
     xmlNodePtr child_node;
 
+    error = _get_request_xml_nodes(client, request, &doc, &root, "ListAllMyBucketsResult");
+    if (error != NULL) {
+        return error;
+    }
+
+    response = g_new0(ds3_get_service_response, 1);
     for (child_node = root->xmlChildrenNode; child_node != NULL; child_node = child_node->next) {
         if (element_equal(child_node, "Buckets") == true) {
-            GArray* buckets_array = _parse_buckets(log, doc, child_node);
-            response->buckets = (ds3_bucket*)buckets_array->data;
+            GPtrArray* buckets_array;
+            error = _parse_bucket_array(client->log, doc, child_node, &buckets_array);
```

Comment:
If seems like we should check to see if an error was returned, and if so break from the loop